rts/Lua/LuaMatrixImpl.cpp

```diff
@@ -0,0 +1,323 @@
+{
+	mat.Translate({x, y, z});
+}
+
+void LuaMatrixImpl::Scale(const float x, const float y, const float z)
+{
+	mat.Scale({x, y, z});
+}
+
+void LuaMatrixImpl::Scale(const float s)
+{
+	mat.Scale({s, s, s});
+}
+
+void LuaMatrixImpl::RotateRad(const float rad, const float x, const float y, const float z)
+{
+	// TODO: Find out why minus and if it is correct.
+	// This code is copy-pasted from develop
+	mat.Rotate(-rad , {x, y, z});
+}
+
+///////////////////////////////////////////////////////////
+
+bool LuaMatrixImpl::UnitMatrix(const unsigned int unitID, const sol::optional<bool> mult, sol::this_state L)
+{
+	return ObjectMatImpl<CUnit>(unitID, mult.value_or(OBJECT_MULT_DEFAULT), L);
+}
+
+bool LuaMatrixImpl::UnitPieceMatrix(const unsigned int unitID, const unsigned int pieceNum, const sol::optional<bool> mult, sol::this_state L)
+{
+	return ObjectPieceMatImpl<CUnit>(unitID, pieceNum, mult.value_or(OBJECT_MULT_DEFAULT), L);
+}
+
+bool LuaMatrixImpl::FeatureMatrix(const unsigned int featureID, const sol::optional<bool> mult, sol::this_state L)
+{
+	return ObjectMatImpl<CFeature>(featureID, mult.value_or(OBJECT_MULT_DEFAULT), L);
+}
+
+bool LuaMatrixImpl::FeaturePieceMatrix(const unsigned int featureID, const unsigned int pieceNum, const sol::optional<bool> mult, sol::this_state L)
+{
+	return ObjectPieceMatImpl<CFeature>(featureID, pieceNum, mult.value_or(OBJECT_MULT_DEFAULT), L);
+}
+
+bool LuaMatrixImpl::ProjectileMatrix(const unsigned int projID, const sol::optional<bool> mult, sol::this_state L)
+{
+	return ObjectMatImpl<CProjectile>(projID, mult.value_or(OBJECT_MULT_DEFAULT), L);
+}
+
+///////////////////////////////////////////////////////////
+
+template<typename TObj>
+bool LuaMatrixImpl::IsObjectVisible(const TObj* obj, sol::this_state L)
+{
+	if (CLuaHandle::GetHandleFullRead(L))
+		return true;
+
+	const int readAllyTeam = CLuaHandle::GetHandleReadAllyTeam(L);
+
+	if ((readAllyTeam < 0) && (readAllyTeam == CEventClient::NoAccessTeam))
+		return false;
+
+	if constexpr(std::is_same<TObj, CProjectile>::value) //CProjectile has no IsInLosForAllyTeam()
+		return losHandler->InLos(obj, readAllyTeam);
+	else //else is needed. Otherwise won't compile
+		return obj->IsInLosForAllyTeam(readAllyTeam);
+}
+
+template<typename TObj>
+const TObj* LuaMatrixImpl::ParseObject(int objID, sol::this_state L)
+{
+	const TObj* obj = nullptr;
+	if constexpr(std::is_same<TObj, CUnit>::value) {
+		obj = unitHandler.GetUnit(objID);
+	}
+	else if constexpr(std::is_same<TObj, CFeature>::value) {
+		obj = featureHandler.GetFeature(objID);
+	}
+	else if constexpr(std::is_same<TObj, CProjectile>::value) {
+		obj = projectileHandler.GetProjectileBySyncedID(objID);
+	}
+
+	if (obj == nullptr)
+		return nullptr;
+
+	if (!IsObjectVisible(obj, L))
+		return nullptr;
+
+	return obj;
+}
+
+const LocalModelPiece* LuaMatrixImpl::ParseObjectConstLocalModelPiece(const CSolidObject* obj, const unsigned int pieceNum)
```

Comment:
this method can be const I think

Reply:
error: static member function 'static const LocalModelPiece* LuaMatrixImpl::ParseObjectConstLocalModelPiece(const CSolidObject*, unsigned int)' cannot have cv-qualifier